### ISSUE TITLE
Implementing `tag011` - the first baseline

### ISSUE THREAD

**Issue by eubinecto** [eubinecto/idiomify#1]
## 최종목표
해당 논문의 literal -> idiomatic style transfer를 구현하는 것이 목표!
<img width="607" alt="image" src="https://user-images.githubusercontent.com/56193069/156719350-631979c4-5187-4a58-ba48-a8dfb802bd21.png">

## Why?
In order to see the advantage we could get from integrating the notion of transparency,
we first need a baseline that does not care about the notion of transparency of idioms. 
And since this is about simple experimentations, we only care about at most nine idioms.

## To-do's
- [x] build a mini dataset: `idioms`
- [x] build a mini dataset: `literal2idiom`
- [x] BART -  `SRCBuilder` 
- [x] BART - `TGTBuilder`
- [x] 데이터 모듈 수정 & 테스트
- [x]  `alpha` 베이스라인 구현하기
- [x]   ainize에서 모델  학습 돌리기
- [x]  `Idiomifier` 구현하기
- [x]  `main_infer` 완성하고, infer pipeline 구축 완료하기 
- [ ]  `streamlit`에  issue_1 배포하기

**Comment by eubinecto** [comment]
## Build a mini dataset: `idioms`

What idioms should I choose? let's first give it a go with  immutable idioms.

three  idioms that are transparent: 
- keep up the good work
- never mind
- now or never

three idioms that are obscure:
- nitty gritty
- the last straw
- let sleeping dogs lie


Let's go for this!

**Comment by eubinecto** [comment]
## build a mini dataset: `literal2idiom`

개발 방향을 바꿨다. 근희님께서 조언해주신대로, 적절한 관용구를 검색하기보단 (나중에 필요할 수는 있겠으나), style을 변경하는 방식을 택했다. 예를 들면 다음과 같은 결과를 기대하는 것:
<img width="1059" alt="image" src="https://user-images.githubusercontent.com/56193069/156703029-adb6d5d9-0327-41f1-8220-9e0281c3b7d5.png">

이 방향이 수업에 사용하기 더 편리할 것이다.

그래서 우리에게 필요한 데이터는  `literal2idiom`  병렬 말뭉치이다.

**Comment by eubinecto** [comment]
## 그렇다면 baseline은?

논문을 읽어보면 copy mechanism, retrieve-delete-replace 등, text style transfer를 구현하기 위해 다양한 트위킹을 시도한 것을 볼 수 있는데, 일단 베이스라인 구축에는 모든 걸 다 구현할 필요가 없다. 

베이스 라인은 Vanilla BART만을 사용해서 한번 구현해보자. 이걸 구현하는 것도 만만치 않을 것이다.

**Comment by eubinecto** [comment]
## BART - `InputsBuilder`

**Comment by eubinecto** [comment]
## 데이터 모듈

아직 테스트, validation, predict는 로더 정의를 하지 않겠다. 일단 빠르게 구현을 해보는 것이 급선무.. !

기존처럼 X, Y로 단순히 나누지 않고, bart의 형식에 맞추어서 src, tgt_r, tgt으로 나누었다
```
    def setup(self, stage: Optional[str] = None):
        """
        setup the builders.
        """
        # --- set up the builders --- #
        # build the datasets
        srcs = SourcesBuilder(self.tokenizer)(self.literal2idiomatic)
        tgts_r = TargetsRightShiftedBuilder(self.tokenizer)(self.literal2idiomatic)
        tgts = TargetsBuilder(self.tokenizer)(self.literal2idiomatic)
        self.dataset = IdiomifyDataset(srcs, tgts_r, tgts)

    def train_dataloader(self) -> DataLoader:
        return DataLoader(self.dataset, batch_size=self.config['batch_size'],
                          shuffle=self.config['shuffle'], num_workers=self.config['num_workers'])

```

**Comment by eubinecto** [comment]
## `Idiomifier` 구현

일단 간단한 베이스라인을 구현하는 것이 목표.  forward와 step은 대충 요롷게 진행이 될듯: 
```python
    def forward(self, srcs: torch.Tensor, tgts_r: torch.Tensor) -> torch.Tensor:
        """
        as for using bart for CG, refer to:
        https://huggingface.co/docs/transformers/model_doc/bart#transformers.BartForQuestionAnswering.forward
        param srcs: (N, 2, L_s)
        param tgts_r: (N, 2, L_t)
        """
        input_ids, attention_mask = srcs[:, 0], srcs[:, 1]
        decoder_input_ids, decoder_attention_mask = tgts_r[:, 0], tgts_r[:, 1]
        outputs = self.bart(input_ids=input_ids,
                            attention_mask=attention_mask,
                            decoder_input_ids=decoder_input_ids,
                            decoder_attention_mask=decoder_attention_mask)
        logits = outputs[0]
        return logits  # (N, L, |V|)

    def predict(self) -> torch.Tensor:
        pass

    def training_step(self, batch: Tuple[torch.Tensor, torch.Tensor, torch.Tensor]) -> dict:
        srcs, tgts_r, tgts = batch  # (N, 2, L_s), (N, 2, L_t), (N, 2, L_t)
        logits = self.forward(srcs, tgts_r)  # -> (N, L, |V|)
        loss = ...
        return {
            "loss": loss
        }

```

**Comment by eubinecto** [comment]
## Streamlit에 배포를 하는건..?
일단 현재로서는 이게 반드시 필요한 건 아니다. 좀 더 성능이 좋아지면, 그때 시도를 해보는 것은 나쁘지 않을 것 같다. 큰 텍스트 박스에 문단 전체를 넣는 방식으로..

**Comment by eubinecto** [comment]
git tag로 버전 관리를 하기로 했다. 공식문서가 꽤나 도움이 된다;
https://git-scm.com/book/en/v2/Git-Basics-Tagging

일단 `tag011` 인 브랜치를 따라 들어가면,  현재 이 버전을 테스트해볼수  있는 상태 그대로 돌아갈 수 있게된다. 

이렇게하는 편이 계속 레거시를 관리하는 것보다 훨~씬 더 스마트하고, 깃을 사용하는 목적일 것 같다!
